### PR TITLE
Aaron/search

### PR DESCRIPTION
**Pull request by aaronnagler**
Updated Pull req (Nov 6, 2023)
Overview:

Implemented and tested the following functions: multi_filter_search_happiness (in happiness.py), get_happiness_by_filter (in happiness_dao.py).

Changes made 
- wrote specification for and implemented multi_filter_search_happiness
- wrote specification for and implemented get_happiness_by_filter
- wrote 8 test cases for multi_filter_search_happiness (which check get_happiness_by_filter)

Test coverage:
- all branches of get_happiness_by_filter are checked
- test case for edge case of get_happiness_by_filter (no filters applied)

Previous Pull Request is Below:
Overview:

Implemented and tested the following functions: get_happiness_from_value_range (in happiness.py) and get_happiness_by_value_range (in happiness_dao.py). Modified miscellaneous test cases in test_happiness.py. 

Changes Made:

- wrote specification for and implemented get_happiness_from_value_range
- wrote specification for and implemented get_happpiness_by_value_range
- wrote two test cases for get_happiness_from_value_range (in function called: test_happiness_from_value_range)
- modified preexisting test cases in the following functions: test_create_happiness_entry, test_overwrite_create_happiness_entry, test_edit_happiness
- created the following functions in the test suite: test_delete_happiness, test_happiness_from_value_range)

Test Coverage:

I wrote two tests for get_happiness_from_value_range, one which had a range between two different numbers, and one case which had a range between the same number. Both cases worked as I expected.

### CODE REVIEW

File: happiness-backend/api/dao/happiness_dao.py [hunk 1]
```diff
@@ -64,7 +66,74 @@ def get_happiness_by_group_count(user_ids, page, n):
         Happiness.timestamp.desc()) \
         .paginate(page=page, per_page=n, error_out=False)
 
+def get_happiness_by_filter(user_id, page, per_page, start, end, low, high, query):
+    """
+    Filters according to the provided arguments. Checks to see what filters to apply. Will not apply the following
+    filters with the following arguments
+     low/high = 11
```
Comment (zachseidner1) [inline happiness-backend/api/dao/happiness_dao.py]: This seems like a very arbitrary choice for a number to not apply the filters. Shouldn't it not apply if low or high is `None`? Why 11? Also will the filters apply if low > high? The docs should specify behavior in this case. Same thing with start > end

File: happiness-backend/api/dao/happiness_dao.py [hunk 2]
```diff
@@ -64,7 +66,74 @@ def get_happiness_by_group_count(user_ids, page, n):
...
+     start/end = None
+     text = "".
+    """
+    if start is not None and low != 11 and query != "":
```
Comment (zachseidner1) [inline happiness-backend/api/dao/happiness_dao.py]: why only check if start is not None and low != 11 without checking end and high? You are making the assumption these values are provided if low and start are provided. This assumption could cause an internal server error if unchecked.

File: happiness-backend/api/dao/happiness_dao.py [hunk 3]
```diff
@@ -64,7 +66,74 @@ def get_happiness_by_group_count(user_ids, page, n):
...
+        return db.paginate(
+            select=
+            (select(Happiness).where(Happiness.timestamp.between(start, end),
+                                     Happiness.user_id == user_id).where(
+                    Happiness.value >= low, Happiness.value <= high,
+                    Happiness.user_id ==
+                    user_id).where(Happiness.comment.like(f"%{query}%")).order_by(Happiness.timestamp.asc())),
```
Comment (zachseidner1) [inline happiness-backend/api/dao/happiness_dao.py]: Why are you chaining together `.where` calls? Couldn't you apply all the filters in a single `where`? [Like this](https://stackoverflow.com/a/72828047)

File: happiness-backend/api/dao/happiness_dao.py [hunk 4]
```diff
@@ -64,7 +66,74 @@ def get_happiness_by_group_count(user_ids, page, n):
...
+            per_page=per_page,
+            page=page
+        )
+    elif start is not None and low != 11:
```
Comment (zachseidner1) [inline happiness-backend/api/dao/happiness_dao.py]: You can definitely rewrite this method to not address each case in a `elif` chain. You could probably do a select statement, then apply `.where` whenever filters are present. For example you start by doing select happiness, then if low and high are there you `.where` on that, then check the next filter and apply `.where` again to the same query object, etc..

File: happiness-backend/api/models/schema.py [hunk 5]
```diff
@@ -223,3 +225,22 @@ class ReadsSchema(ma.Schema):
...
+    low = ma.Int()
+    high = ma.Int()
+
+
+class HappinessMultiFilterSchema(ma.Schema):
+    user_id = ma.Int()
+    page = ma.Int()
+    count = ma.Int()
+    low = ma.Int(missing=11)
+    high = ma.Int(missing=11)
```
Comment (zachseidner1) [inline happiness-backend/api/models/schema.py]: why 11 😠

File: happiness-backend/api/models/schema.py [hunk 6]
```diff
@@ -223,3 +225,22 @@ class ReadsSchema(ma.Schema):
...
+    start = ma.Str(missing=None)
+    end = ma.Str(missing=None)
```
Comment (zachseidner1) [inline happiness-backend/api/models/schema.py]: missing=None is not required here since `.get` will return None if the value is not present.

Comment (zachseidner1) [inline happiness-backend/api/models/schema.py]: Also look into refactoring this to use `ma.Date()` I saw it in Jonathan's code so feel free to ask him.

File: happiness-backend/tests/test_happiness.py
Comment (zachseidner1) [inline happiness-backend/tests/test_happiness.py:1]: Nice job cleaning this up!

File: happiness-frontend/package-lock.json
Comment (zachseidner1) [inline happiness-frontend/package-lock.json:1]: Why is this file different? I don't think this change should be a part of your PR.

File: happiness-backend/tests/test_happiness.py [hunk 7]
```diff
@@ -427,3 +406,109 @@ def test_discussion_comments(init_client):
     comments = comment_access_u3.json
     assert len(comments) == 2
     assert comments[0]['author']['id'] == 1 and comments[1]['author']['id'] == 3
+
+
+def test_happiness_search(init_client):
```
Comment (zachseidner1) [inline happiness-backend/tests/test_happiness.py:411]: Just as you did for Happiness, split this into separate, smaller unit test cases. For example one for searching with value, one for searching with dates, one for both, etc.. This way if the test fails it will be easier to know which is going wrong.

**Review by zachseidner1 — CHANGES_REQUESTED** [review]
LGTM (looks garbage to me)

jk jk jk jk jk jk!!! Aaron congrats on your first PR. You definitely have a good head on your shoulders since you were able to pick all of this up and get adjusted to the codebase relatively quickly, especially with no backend experience. It seems like you are pretty comfortable with what is going on at this point, so nice job! You definitely need to fix up some things I left comments on, but overall for your first PR this is some good work and I'm glad you're helping out and were able to pick up on what's going on with the codebase.

File: happiness-backend/api/dao/happiness_dao.py (hunk 4 above)
Comment (jonathanjma) [reply happiness-backend/api/dao/happiness_dao.py]: I agree :-)

File: happiness-backend/api/dao/happiness_dao.py (hunk 3 above)
Comment (jonathanjma) [reply happiness-backend/api/dao/happiness_dao.py]: ‼️

**Review by jonathanjma — CHANGES_REQUESTED** [review]

File: happiness-backend/api/dao/happiness_dao.py (hunk 1 above)
Comment (aaronnagler) [reply happiness-backend/api/dao/happiness_dao.py]: i chose 11 bc happiness entries cannot be a value 11, but i understand its arbritrary, do you think chaging the default value to None would be better? i also agree about the specified behavior, so im changing it so itll raise exceptions if low>high or start ">" end, thanks!

File: happiness-backend/api/dao/happiness_dao.py (hunk 2 above)
Comment (aaronnagler) [reply happiness-backend/api/dao/happiness_dao.py]: good point

File: happiness-backend/api/dao/happiness_dao.py (hunk 3 above)
Comment (aaronnagler) [reply happiness-backend/api/dao/happiness_dao.py]: lol i didn't know that exists indeed this would be better

File: happiness-backend/api/dao/happiness_dao.py (hunk 4 above)
Comment (aaronnagler) [reply happiness-backend/api/dao/happiness_dao.py]: i think i understand this point but i don't really undestand exactly what you mean or how to do this. correct me if im wrong, but from my current understanding, instead of putting my boolean expressions which are checking to see which arguments are "real" on the outside of the select statement, i should instead put this logic inside the select statement, like this
 select=
        (select(Happiness).where( (if start is not None : 
            
            (Happiness.timestamp.between(start, end)) & if low != 11: 
Happiness.value >= low, Happiness.value <= high)
.order_by(Happiness.timestamp.asc())),
        per_page=per_page,
        page=page
    )
or something like this, ignore my bad indenting sorry, but i don't really know how to format this bc where is expecting filters, not boolean logic to my knowledge, and if we put the boolean logic outside of the where statement, or outside of the select statement, then aren't we back to the original logic i had, needing to have multiple select statements?

Comment (aaronnagler) [reply happiness-backend/api/dao/happiness_dao.py]: i think i understand this point but i don't really undestand exactly what you mean or how to do this. correct me if im wrong, but from my current understanding, instead of putting my boolean expressions which are checking to see which arguments are "real" on the outside of the select statement, i should instead put this logic inside the select statement, like this
 select=
        (select(Happiness).where( (if start is not None : 
            
            (Happiness.timestamp.between(start, end)) & if low != 11: 
Happiness.value >= low, Happiness.value <= high)
.order_by(Happiness.timestamp.asc())),
        per_page=per_page,
        page=page
    )
or something like this, ignore my bad indenting sorry, but i don't really know how to format this bc where is expecting filters, not boolean logic to my knowledge, and if we put the boolean logic outside of the where statement, or outside of the select statement, then aren't we back to the original logic i had, needing to have multiple select statements?

File: happiness-backend/api/models/schema.py (hunk 5 above)
Comment (aaronnagler) [reply happiness-backend/api/models/schema.py]: can i change it to None in the schema for the optional argument?
heres what the schema looks like at the moment, should i just make all the optional arguments none, like including the text one which is currently the empty string? i assume that would be fine to do right?
  user_id = ma.Int()
    page = ma.Int()
    count = ma.Int()
    low = ma.Int(missing=11)
    high = ma.Int(missing=11)
    start = ma.Str(missing=None)
    end = ma.Str(missing=None)
    text = ma.Str(missing="")

File: happiness-backend/api/models/schema.py (hunk 6 above)
Comment (aaronnagler) [reply happiness-backend/api/models/schema.py]: thanks

File: happiness-frontend/package-lock.json
Comment (aaronnagler) [reply happiness-frontend/package-lock.json:1]: interesting yeah i have no idea what file that is either lol, is there a way i can not make that part of my pr, do i not add all my files in my git add . or something?

File: happiness-backend/tests/test_happiness.py (hunk 7 above)
Comment (aaronnagler) [reply happiness-backend/tests/test_happiness.py:411]: makes sense but i have one concern which is for all of my tests on search they currently all search within a set of happiness entries that i created with a bunch of post requests from lines 413 to line 453, basically establishing a bunch of entries for me to test filtering through. If i seperate all the searches into multiple test cases, will i need to copy these lines into all of those test cases? i assume there is a better way than to do that, maybe I could call those test cases within one larger test case, but if i did that would the program treat those calls as test cases or just helper functions?

**Review by jonathanjma — APPROVED** [review]
Looks good to me! Great work on your first PR :-)

**Review by zachseidner1 — APPROVED** [review]
LGTM (let's get this merged)!
